src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -964,6 +1085,64 @@ public DocumentId AddTextContainer(SourceTextContainer textContainer, string ful
                 return documentId;
             }
 
+            public void AddDynamicFile(IDynamicFileInfoProvider fileInfoProvider, DynamicFileInfo fileInfo, ImmutableArray<string> folders)
+            {
+                var documentInfo = CreateDocumentInfoFromFileInfo(fileInfo, folders.NullToEmpty());
+                var documentId = documentInfo.Id;
```

Comment:
feels very weird to do this here before we've actually done our argument checking.